benchmark/common.js

```diff
@@ -88,52 +88,157 @@ Benchmark.prototype._queue = function(options) {
-      'benchmarks. Check benchmark/README.md for further instructions.');
-    process.exit(1);
-  }
+function AutocannonBenchmarker() {
+  const autocannon_exe = process.platform === 'win32'
+                         ? 'autocannon.cmd'
+                         : 'autocannon';
+  this.present = function() {
+    var result = child_process.spawnSync(autocannon_exe, ['-h']);
+    if (result.error && result.error.code === 'ENOENT')
+      return false;
+    else
+      return true;
+  };
+  this.create = function(path, duration, connections) {
+    const args = ['-d', duration, '-c', connections, '-j', '-n',
+                  'http://127.0.0.1:' + exports.PORT + path ];
+    var child = child_process.spawn(autocannon_exe, args);
+    child.stdout.setEncoding('utf8');
+    return child;
+  };
+  this.processResults = function(output) {
+    let result;
+    try {
+      result = JSON.parse(output);
+    } catch (err) {
+      // Do nothing, let next line handle this
+    }
+    if (!result || !result.requests || !result.requests.average) {
+      return undefined;
+    } else {
+      return result.requests.average;
+    }
+  };
 }
 
-// benchmark an http server.
-const WRK_REGEXP = /Requests\/sec:[ \t]+([0-9\.]+)/;
-Benchmark.prototype.http = function(urlPath, args, cb) {
-  hasWrk();
-  const self = this;
-
-  const urlFull = 'http://127.0.0.1:' + exports.PORT + urlPath;
-  args = args.concat(urlFull);
-
-  const childStart = process.hrtime();
-  const child = child_process.spawn('wrk', args);
-  child.stderr.pipe(process.stderr);
+function WrkBenchmarker() {
+  this.present = function() {
+    var result = child_process.spawnSync('wrk', ['-h']);
+    if (result.error && result.error.code === 'ENOENT')
+      return false;
+    else
+      return true;
+  };
+  this.create = function(path, duration, connections) {
+    const args = ['-d', duration, '-c', connections, '-t', 8,
+                  'http://127.0.0.1:' + exports.PORT + path ];
+    var child = child_process.spawn('wrk', args);
+    child.stdout.setEncoding('utf8');
+    child.stderr.pipe(process.stderr);
+    return child;
+  };
+  const regexp = /Requests\/sec:[ \t]+([0-9\.]+)/;
+  this.processResults = function(output) {
+    const match = output.match(regexp);
+    const result = match && +match[1];
+    if (!result)
+      return undefined;
+    else
+      return result;
+  };
+}
 
-  // Collect stdout
-  let stdout = '';
-  child.stdout.on('data', (chunk) => stdout += chunk.toString());
+const HTTPBenchmarkers = {
+  autocannon: new AutocannonBenchmarker(),
+  wrk: new WrkBenchmarker()
+};
 
-  child.once('close', function(code) {
-    const elapsed = process.hrtime(childStart);
-    if (cb) cb(code);
+// Benchmark an http server.
+Benchmark.prototype.http = function(urlPath, duration, connections, cb) {
+  const self = this;
+  duration = 1;
+
+  const picked_benchmarker = process.env.NODE_HTTP_BENCHMARKER ||
```

Comment:
For the record: I have done a crude interaction analysis, it turns out that the benchmarker does affect performances when looking at a specific set of parameters:

```
  c chunks length    type improvement significant      p.value
  50      0      4  buffer     -7.91 %         *** 1.322680e-09
  50      0      4   bytes     17.84 %         *** 4.045045e-21
  50      0   1024  buffer    -20.69 %         *** 9.271728e-27
  50      0   1024   bytes    -22.11 %         *** 1.515380e-20
  50      0 102400  buffer    -30.34 %         *** 5.798797e-49
  50      0 102400   bytes      3.83 %         *** 2.321257e-11
  50      1      4  buffer    -12.34 %         *** 3.224916e-16
  50      1      4   bytes    -10.43 %         *** 5.706192e-13
  50      1   1024  buffer    -15.65 %         *** 2.850410e-07
  50      1   1024   bytes     22.22 %         *** 1.248468e-39
  50      1 102400  buffer    -31.47 %         *** 4.988408e-46
  50      1 102400   bytes     -0.63 %          ** 6.046574e-03
  50      4      4  buffer     38.50 %         *** 2.366325e-52
  50      4      4   bytes     29.70 %         *** 2.854201e-27
  50      4   1024  buffer     27.75 %         *** 6.241389e-36
  50      4   1024   bytes     64.73 %         *** 4.793425e-22
  50      4 102400  buffer     -8.18 %         *** 7.693781e-26
  50      4 102400   bytes      3.94 %         *** 4.965380e-06
 500      0      4  buffer      9.04 %         *** 7.695347e-18
 500      0      4   bytes     17.01 %         *** 3.182603e-34
 500      0   1024  buffer    -14.02 %         *** 2.095410e-33
 500      0   1024   bytes    -19.36 %         *** 1.450384e-32
 500      0 102400  buffer    -38.56 %         *** 2.547833e-67
 500      0 102400   bytes      4.35 %         *** 7.185082e-20
 500      1      4  buffer      8.00 %         *** 1.383958e-21
 500      1      4   bytes      5.21 %         *** 7.024325e-16
 500      1   1024  buffer     -9.36 %         *** 2.184297e-20
 500      1   1024   bytes      6.08 %         *** 7.288844e-14
 500      1 102400  buffer    -36.23 %         *** 1.578285e-65
 500      1 102400   bytes      1.00 %         *** 2.875669e-04
 500      4      4  buffer     19.72 %         *** 7.854005e-49
 500      4      4   bytes     44.20 %         *** 1.691548e-35
 500      4   1024  buffer     19.78 %         *** 1.023951e-38
 500      4   1024   bytes     34.85 %         *** 3.279581e-39
 500      4 102400  buffer     -9.27 %         *** 1.401027e-28
 500      4 102400   bytes      6.78 %         *** 1.708493e-32
```
_relative performances improvement of using autocannon._

however this doesn't say anything about the benchmarks ability to benchmark an optimization proposal. It just means that the benchmarkers aren't equally performant in all aspects. Really it just optimization suggests for the benchmarker implementers :)

_Anyway it was just for the record._